src/models/tests/gamma_poisson_model_test.py

```diff
@@ -0,0 +1,198 @@
+        )
+
+    def test_knreach(self):
+        gpm = GammaPoissonModel([ReachPoint([], [])])
+        self.assertAlmostEqual(gpm._knreach(0, 1, 1, 2, 1.0, 1.0), 1 / 8)
+        self.assertAlmostEqual(gpm._knreach(1, 1, 1, 2, 1.0, 1.0), 1 / 8)
+        self.assertAlmostEqual(gpm._knreach(2, 1, 1, 2, 1.0, 1.0), 0)
+        self.assertAlmostEqual(gpm._knreach(1, 1, 1, 2, 1.0, 2.0), 1.0 / 9.0)
+        self.assertAlmostEqual(gpm._knreach(1, [1], 1, 2, 1.0, 1.0)[0], 1 / 8)
+        self.assertAlmostEqual(gpm._knreach(1, [1, 2], 1, 2, 1.0, 1.0)[0], 1 / 8)
+        self.assertAlmostEqual(gpm._knreach(1, [1, 2], 1, 2, 1.0, 1.0)[1], 1 / 16)
+        self.assertAlmostEqual(
+            gpm._knreach(1, 3, 1, 3, 1.0, 1.0), 3 * (1 / 3) * (2 / 3) ** 2 * (1 / 16)
+        )
+
+    def test_kreach(self):
+        gpm = GammaPoissonModel([ReachPoint([], [])])
+        self.assertAlmostEqual(gpm._kreach([0], 1, 2, 1, 1)[0], 2 / 3)
+        self.assertAlmostEqual(gpm._kreach([0], 1, 3, 1, 1)[0], 3 / 4)
+        self.assertAlmostEqual(gpm._kreach([1], 1, 3, 1, 1)[0], 3 / 16)
+        self.assertAlmostEqual(gpm._kreach([0, 1, 2], 1, 3, 1, 1)[0], 3 / 4)
+        self.assertAlmostEqual(gpm._kreach([0, 1, 2], 1, 3, 1, 1)[1], 3 / 16)
+        self.assertAlmostEqual(gpm._kreach([0, 1, 2], 1, 3, 1, 1)[2], 3 / 64)
+
+    def test_expected_impressions(self):
+        gpm = GammaPoissonModel([ReachPoint([], [])])
+        self.assertAlmostEqual(gpm._expected_impressions(1, 1, 1), 1.0)
+        self.assertAlmostEqual(gpm._expected_impressions(2, 1, 1), 2.0)
+        self.assertAlmostEqual(gpm._expected_impressions(1, 2, 1), 2)
+        self.assertAlmostEqual(gpm._expected_impressions(1, 1, 2), 0.5)
+
+    def test_expected_histogram(self):
+        gpm = GammaPoissonModel([ReachPoint([], [])])
+        h_actual = gpm._expected_histogram(4, 12, 16, 1, 1, max_freq=3)
+        self.assertLen(h_actual, 3)
+        self.assertAlmostEqual(h_actual[0], 3)
+        self.assertAlmostEqual(h_actual[1], 3 / 4.0)
+        self.assertAlmostEqual(h_actual[2], 3 / 16.0)
+
+    def test_feasible_point(self):
+        gpm = GammaPoissonModel([ReachPoint([], [])])
+        alpha, beta, I = gpm._feasible_point([14, 8, 4, 2])
+        self.assertAlmostEqual(alpha, 1.0)
+        self.assertAlmostEqual(beta, 56 / 500)
+        self.assertAlmostEqual(I, 500)
+
+    def test_fit_histogram_chi2_distance(self):
+        gpm = GammaPoissonModel([ReachPoint([], [])])
+        I0, N0, alpha0, beta0 = 20000, 10000, 5, 0.5
+        h_actual = gpm._expected_histogram(I0, I0 / 0.1, N0, alpha0, beta0)
+        Imax, N, alpha, beta = gpm._fit_histogram_chi2_distance(h_actual)
+        Ih = np.sum([h_actual[i] * (i + 1) for i in range(len(h_actual))])
+        h_predicted = gpm._expected_histogram(Ih, Imax, N, alpha, beta)
+        for i in range(len(h_actual)):
+            self.assertTrue(
+                (h_actual[i] - h_predicted[i]) ** 2 < 15,
```

Comment:
15 is empirical?

Reply:
This was an oversight.  It should have been 1.